### ISSUE TITLE
Auto traits do not get de-duplicated properly when principal trait is auto trait

### ISSUE THREAD

**Issue by alexreg** [rust-lang/rust#57057]
This was an issue that came up during work on https://github.com/rust-lang/rust/pull/55994.

Basically, a bug in the de-duplication code means that `dyn Send` is a different type from `dyn Send + Send`, and likewise `dyn Sync + Send` is different from `dyn Sync + Send + Sync`. However, this clearly should not be the case, and indeed `dyn Send + Send` is the same type as `dyn Send + Send + Send`. (Note, this is only tangentially related to the lint https://github.com/rust-lang/rust/issues/56522.)

Apart from potentially surprising behaviour, the issue with the above lies with trait impls and coherence. I am unaware of a specific ICE that can arise because of it, but it is probably UB in general.

```rust
trait Foo {}

impl Foo for dyn Send {}

impl Foo for dyn Send + Send {}
```

[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=2763112d608c2c04d39dd80549b60799)

```rust
trait Foo {}

impl Foo for dyn Send + Send {}

impl Foo for dyn Send + Send + Send {}
```

[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d88e26708a2ed6409c7d405270eb05e0)

CC @nikomatsakis @arielb1

**Comment by alexreg** [comment]
Note, I intend to fix this in https://github.com/rust-lang/rust/pull/55994 of course, but this services as a reference point at least, and will be cited in code.

**Comment by arielb1** [comment]
This will be fixed in #56873
